compiler/rustc_middle/src/mir/interpret/pointer.rs

```diff
@@ -83,55 +83,74 @@ pub trait PointerArithmetic: HasDataLayout {
 
 impl<T: HasDataLayout> PointerArithmetic for T {}
 
+/// This trait abstracts over the kind of provenance that is associated with a `Pointer`. It is
+/// mostly opaque; the `Machine` trait extends it with some more operations that also have access to
+/// some global state.
+pub trait Provenance: Copy {
+    /// Says whether the `offset` field of `Pointer`s with this provenance is the actual physical address.
+    /// If `true, ptr-to-int casts work by simply discarding the provenance.
+    /// If `false`, ptr-to-int casts are not supported.
+    const OFFSET_IS_ADDR: bool;
+
+    /// Determines how a pointer should be printed.
+    fn fmt(ptr: &Pointer<Self>, f: &mut fmt::Formatter<'_>) -> fmt::Result
+    where
+        Self: Sized;
+
+    /// "Erasing" a tag converts it to the default tag type if possible. Used only for formatting purposes!
+    fn erase_for_fmt(self) -> AllocId;
+}
+
+impl Provenance for AllocId {
+    // With the `AllocId` as provenance, the `offset` is interpreted *relative to the allocation*,
+    // so ptr-to-int casts are not possible (since we do not know the global physical offset).
+    const OFFSET_IS_ADDR: bool = false;
+
+    fn fmt(ptr: &Pointer<Self>, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Comment:
Why is there a custom formatting function? Couldn't this be an impl on `Pointer<AllocId>` directly, allowing all aggregates containing `Pointer` to derive their `Debug` impls?

Reply:
I don't see how. The `Tag` type affects how the `Pointer` itself is rendered:
- `Pointer<AllocId>`: `alloc123+0x13`
- `Pointer<Tag>`: `0x12defg60[alloc123]<Stacked Borrows info>`

Reply:
>  I'm mainly unhappy about the roundabout formatting stuff that requires one to essentially write the derived impl by hand.

Yeah I am unhappy about that as well.^^ Open to suggestions for how to do this better.

Reply:
can miri implement Display for Pointer<Tag> since Tag is a miri-internal type?

Reply:
Doesn't that only work for "fundamental" types?

Reply:
Oh also that would not help -- the generic code in `rustc_mir::interpret` needs to know for sure that it can debug-print pointers. So we need something generic.

Is it possible to add `Pointer<Self::PointerTag>: Debug` as a requirement to the `Machine` trait?^^

Reply:
> Is it possible to add `Pointer<Self::PointerTag>: Debug` as a requirement to the `Machine` trait?^^

Not in a useful way (you'd have to repeat the bound at every usage site of `Machine`).

*but* you could keep your `fmt` function and only implement `Debug` manually on `Pointer<Tag: Provenance>` and then you should be able to derive `Debug` everywhere else, as long as you add the `Tag: Provenance` bound to all the types instead of only on the impls. It's not nice, but nicer than doing manual trivial impls everywhere I think?

Reply:
thinking about this again, it's not too bad, so we can merge and figure out improvements later